cli/src/main/kotlin/com/bazel_diff/bazel/BazelClient.kt

```diff
@@ -6,15 +6,33 @@ import org.koin.core.component.KoinComponent
 import org.koin.core.component.inject
 import java.util.*
 
-class BazelClient(private val fineGrainedHashExternalRepos: Set<String>) : KoinComponent {
+class BazelClient(private val useCquery: Boolean, private val fineGrainedHashExternalRepos: Set<String>) : KoinComponent {
     private val logger: Logger by inject()
     private val queryService: BazelQueryService by inject()
 
     suspend fun queryAllTargets(): List<BazelTarget> {
         val queryEpoch = Calendar.getInstance().getTimeInMillis()
 
-        val query = listOf("//external:all-targets", "//...:all-targets") + fineGrainedHashExternalRepos.map { "@$it//...:all-targets" }
-        val targets = queryService.query(query.joinToString(" + ") { "'$it'" })
+        val repoTargetsQuery = listOf("//external:all-targets")
+        val targets = if (useCquery) {
+            // Explicitly listing external repos here sometimes causes issues mentioned at
+            // https://bazel.build/query/cquery#recursive-target-patterns. Hence, we query all dependencies with `deps`
+            // instead. However, we still need to append all "//external:*" targets because fine-grained hash
+            // computation depends on hashing of source files in external repos as well, which is limited to repos
+            // explicitly mentioned in `fineGrainedHashExternalRepos` flag. Therefore, for any repos not mentioned there
+            // we are still relying on the repo-generation target under `//external` to compute the hash.
+            //
+            // In addition, we must include all source dependencies in this query in order for them to show up in
+            // `configuredRuleInput`. Hence, one must not filter them out with `kind(rule, deps(..))`. However, these
+            // source targets are omitted inside BazelQueryService with the custom starlark function used to print
+            // labels.
+            (queryService.query("deps(//...:all-targets)", useCquery = true) +
+                queryService.query(repoTargetsQuery.joinToString(" + ") { "'$it'" }))
+                .distinctBy { it.rule.name }
+        } else {
+            val buildTargetsQuery = listOf("//...:all-targets") + fineGrainedHashExternalRepos.map { "@$it//...:all-targets" }
```

Comment:
What was the reason removing `//external:all-targets` from `query` branch? Without those repo generation rules, how would "coarse grained" hashes get computed?

Reply:
#263 is a related change where I removed rule input transformation for `query`.